File: problem_20.cpp
#include <chrono>
#include <iostream>
#include <cmath>
#include <cstring>
#include <vector>
#include <array>

std::vector<unsigned char> targetNum = {0, 0, 1};
const unsigned char digitMaxVal = 10;

std::vector<unsigned char> multiply_strings (std::vector<unsigned char> f1, std::vector<unsigned char> f2) 
{
    std::vector<unsigned char> res;
    res.resize(f1.size() + f2.size() + 1);
    memset(res.data(), 0, res.size() * sizeof(unsigned char));
    for (int i = 0; i < f1.size(); i++) 
    {
        for (int j = 0; j < f2.size(); j++) 
        {
            int resPos = i + j;
            int digitMult = f1.at(i) * f2.at(j);
            res.at(resPos) += (digitMult) % digitMaxVal;
            if (digitMult >= digitMaxVal) 
            {
                res.at(resPos + 1) += int(digitMult / digitMaxVal);
            }
            for (int k = resPos; k < res.size(); k++) 
            {
                int rip = int(res.at(k) / digitMaxVal);
                if (rip >= 1) 
                {
                    res.at(k) -= rip * digitMaxVal;
                    res.at(k + 1) += rip;
                }
                else 
                {
                    break;
                }
            }
        }
    }
    while (res.size() > 0 && res.at(res.size() - 1) == 0) 
    {
        res.resize(res.size() - 1);
    }
    return res;
}

std::vector<unsigned char> decrement_string (std::vector<unsigned char> n)
{
    for (int i = 0; i < n.size(); i++)
    {
        // I don't need to iterate more than this; decrement and if needed truncate the length
        if (n.at(i) > 0) 
        {
            n.at(i) -= 1;
            if (n.at(i) == 0 && i > 0 && i == n.size() - 1) 
            {
                n.resize(n.size() - 1);
            }
            return n;
        }
        else 
        {
            n.at(i) = 9;
        }
    }
    return n;
}

std::vector<unsigned char> factorial_strings (std::vector<unsigned char> n) 
{
    if (n.size() == 1 && n.at(0) <= 1) 
    {
        return n;
    }
    return multiply_strings(n, factorial_strings(decrement_string(n)));
}

int main() 
{
    auto startTm = std::chrono::high_resolution_clock::now();

    int runningSum = 0;
    std::vector<unsigned char> res = factorial_strings(targetNum);
    for (int i = 0; i < res.size(); i++) 
    {
        runningSum += (int) res.at(i);
        std::cout << int(res.at(res.size() - 1 - i));
    }
    std::cout << std::endl;
    
    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Solution: " << runningSum << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_20.rs
use std::time::{Instant};

const DIGIT_MAX_VAL: u8 = 10;

fn multiply_strings (f1: &Vec<u8>, f2: &Vec<u8>) -> Vec<u8>
{
    let mut res: Vec<u8> = vec![];
    res.resize(f1.len() + f2.len() + 1, 0);
    for i in 0..f1.len()
    {
        for j in 0..f2.len()
        {
            let res_pos: usize = i + j;
            let digit_mult: u8 = f1[i] * f2[j];
            res[res_pos] = res[res_pos] + (digit_mult) % DIGIT_MAX_VAL;
            if digit_mult >= DIGIT_MAX_VAL 
            {
                res[res_pos + 1] = res[res_pos + 1] + (digit_mult / DIGIT_MAX_VAL);
            }
            for k in res_pos..res.len()
            {
                let rip: u8 = res[k] / DIGIT_MAX_VAL;
                if rip >= 1 
                {
                    res[k] = res[k] - rip * DIGIT_MAX_VAL;
                    res[k + 1] = res[k + 1] + rip;
                }
                else 
                {
                    break;
                }
            }
        }
    }
    while res.len() > 0 && res[res.len() - 1] == 0 
    {
        res.resize(res.len() - 1, 0);
    }
    return res;
}

fn decrement_string (n: &Vec<u8>) -> Vec<u8>
{
    let mut res = n.clone();
    for i in 0..res.len()
    {
        // I don't need to iterate more than this; decrement and if needed truncate the length
        if res[i] > 0 
        {
            res[i] = res[i] - 1;
            if res[i] == 0 && i > 0 && i == res.len() - 1 
            {
                res.resize(res.len() - 1, 0);
            }
            return res;
        }
        else 
        {
            res[i] = DIGIT_MAX_VAL - 1;
        }
    }
    return res;
}

fn factorial_strings (n: Vec<u8>) -> Vec<u8>
{
    if n.len() == 1 && n[0] <= 1 
    {
        return n;
    }
    return multiply_strings(&n, &factorial_strings(decrement_string(&n)));
}


fn main () 
{
    let start_tm = Instant::now();

    let target_num: Vec<u8> = vec![0, 0, 1];
    let res: Vec<u8> = factorial_strings(target_num);
    let running_sum: usize = res.iter().fold(0, |acc, x| acc + (*x as usize));

    println!("Result: {}", running_sum);
    let duration = start_tm.elapsed();
    println!("Timing: {}", duration.as_nanos());
}
